libs/datamodel/core/src/walkers.rs

```diff
@@ -105,104 +108,104 @@ impl<'a> ModelRef<'a> {
         match &self.field.field_type {
-            FieldType::Enum(name) => TypeRef::Enum(EnumRef {
+            FieldType::Enum(name) => TypeWalker::Enum(EnumWalker {
                 datamodel: self.datamodel,
                 r#enum: self.datamodel.find_enum(name).unwrap(),
             }),
-            FieldType::Base(scalar_type, _) => TypeRef::Base(*scalar_type),
-            _ => TypeRef::Other,
+            FieldType::Base(scalar_type, _) => TypeWalker::Base(*scalar_type),
+            _ => TypeWalker::Other,
         }
     }
 
-    pub(super) fn is_id(&self) -> bool {
+    pub fn is_id(&self) -> bool {
         self.field.is_id
     }
 
-    pub(super) fn is_required(&self) -> bool {
+    pub fn is_required(&self) -> bool {
         match self.arity() {
             FieldArity::Required => true,
             _ => false,
         }
     }
 
-    pub(super) fn is_unique(&self) -> bool {
+    pub fn is_unique(&self) -> bool {
         self.field.is_unique
     }
 
-    pub(super) fn model(&self) -> ModelRef<'a> {
-        ModelRef {
+    pub fn model(&self) -> ModelWalker<'a> {
+        ModelWalker {
             model: self.model,
             datamodel: self.datamodel,
         }
     }
 
-    pub(super) fn name(&self) -> &'a str {
+    pub fn name(&self) -> &'a str {
         &self.field.name
     }
 }
 
 #[derive(Debug)]
-pub(super) enum TypeRef<'a> {
-    Enum(EnumRef<'a>),
+pub enum TypeWalker<'a> {
+    Enum(EnumWalker<'a>),
     Base(ScalarType),
     Other,
 }
 
-impl<'a> TypeRef<'a> {
-    pub(super) fn as_enum(&self) -> Option<EnumRef<'a>> {
+impl<'a> TypeWalker<'a> {
+    pub fn as_enum(&self) -> Option<EnumWalker<'a>> {
         match self {
-            TypeRef::Enum(r) => Some(*r),
+            TypeWalker::Enum(r) => Some(*r),
             _ => None,
         }
     }
 
-    pub(super) fn is_json(&self) -> bool {
-        matches!(self, TypeRef::Base(ScalarType::Json))
+    pub fn is_json(&self) -> bool {
+        matches!(self, TypeWalker::Base(ScalarType::Json))
     }
 }
 
 #[derive(Debug)]
-pub(super) struct RelationFieldRef<'a> {
+pub struct RelationFieldWalker<'a> {
     datamodel: &'a Datamodel,
     model: &'a Model,
     field: &'a RelationField,
 }
 
-impl<'a> RelationFieldRef<'a> {
-    pub(super) fn arity(&self) -> FieldArity {
+impl<'a> RelationFieldWalker<'a> {
+    pub fn arity(&self) -> FieldArity {
         self.field.arity
     }
 
-    pub(crate) fn is_one_to_one(&self) -> bool {
+    pub fn is_one_to_one(&self) -> bool {
         self.field.is_singular() && self.opposite_side().map(|rel| rel.field.is_singular()).unwrap_or(false)
     }
 
-    pub(crate) fn is_virtual(&self) -> bool {
+    pub fn is_virtual(&self) -> bool {
         self.field.relation_info.fields.is_empty()
     }
 
-    pub(crate) fn opposite_side(&self) -> Option<RelationFieldRef<'a>> {
+    pub fn opposite_side(&self) -> Option<RelationFieldWalker<'a>> {
         self.referenced_model_ref().relation_fields().find(|relation_field| {
```

Comment:
we have self.datamodel.find_related_field_bang(self.field), which always returns the related field. I think when you are operating on the datamodel there should always be an opposite field right? Maybe you can use that and we can cut down on the variants of logic of identifying the opposite.

Reply:
Can you get back to the model of the related field from the return type `find_related_field_bang()`? That would be my only concern. I'll check.

Reply:
You'd have that in the .to field of the initial field at least.

Reply:
So yes, you lose the relation to the model in the other implementation. OTOH you can get the model from `.related_model()`, but it's more work if you want to re-build a walker type from that.

Reply:
I'll make the return type non-optional, you're right it should always be there.

Reply:
ah wait, I found a way to share the logic between the two

Reply:
Hmm, that's unfortunate. My main concern is that I do not want multiple versions of the below logic spread around. Maybe you can just extract and share that separately?

```
relation_field.relation_name() == self.relation_name()
                    && relation_field.referenced_model().name.as_str() == self.model.name
                    // This is to differentiate the opposite field from self in the self relation case.
                    && relation_field.field.relation_info.to_fields != self.field.relation_info.to_fields
                    && relation_field.field.relation_info.fields != self.field.relation_info.fields
```

Reply:
Hmm the logic is already different:

```rust
    pub fn opposite_side(&self) -> RelationFieldWalker<'a> {
        self.referenced_model_ref().relation_fields().find(|relation_field| {
            relation_field.relation_name() == self.relation_name()
                    && relation_field.referenced_model().name.as_str() == self.model.name
                    // This is to differentiate the opposite field from self in the self relation case.
                    && relation_field.field.relation_info.to_fields != self.field.relation_info.to_fields
                    && relation_field.field.relation_info.fields != self.field.relation_info.fields
        })
    }
```

vs

```rust
    /// Finds a relation field related to a relation info
    pub fn find_related_field_for_info(&self, info: &RelationInfo, exclude: &str) -> Option<&RelationField> {
        self.find_model(&info.to)
            .expect("The model referred to by a RelationInfo should always exist.")
            .relation_fields()
            .find(|f| {
                f.relation_info.name == info.name
                    && (f.relation_info.to != info.to ||
                    // This is to differentiate the opposite field from self in the self relation case.
                    f.name != exclude)
            })
    }

```

Reply:
We should definitely reconcile that.

Reply:
The other method uses this as base for example:
```
 f.relation_info.name == info.name
                    && (f.relation_info.to != info.to ||
                    // This is to differentiate the opposite field from self in the self relation case.
                    f.name != exclude)
```

Reply:
I'm going to try using find_related_field_for_info() as a base for the walker version.

Reply:
The upper one for me errored in the case of m2m self relations if I remember correctly. Because for these both relation fields will have fields: vec![] and to_fields: vec![id].

Reply:
This is a test cased I added because the field comparison logic errored for me:

```rust
#[test]
fn implicit_many_to_many_relation() {
    let datamodel = convert(
        r#"model Post {
                    id         String @id @default(cuid())
                    identifier Int?   @unique
                    related    Post[] @relation(name: "RelatedPosts")
                    parents   Post[] @relation(name: "RelatedPosts")
                  }
                  "#,
    );

    let post = datamodel.assert_model("Post");
    post.assert_relation_field("related");

    post.assert_relation_field("parents");
}
```

Reply:
I just did that (use find_related_field_bang() to share the logic). I have enough confidence in the ME test suite to catch problems with that, so I think it's good now :)